transports/quic/src/socket.rs

```diff
@@ -0,0 +1,134 @@
+
+//! I/O for libp2p-quic.
+//!
+//! This provides a central location for socket I/O, and logs all incoming and outgoing packets.
+
+use async_std::net::UdpSocket;
+use log::{trace, warn};
+use quinn_proto::Transmit;
+use std::{future::Future, io::Result, task::Context, task::Poll};
+
+/// A pending packet for libp2p-quic
+#[derive(Debug, Default)]
+pub(crate) struct Pending {
+    pending: Option<Transmit>,
+}
+
+/// A socket wrapper for libp2p-quic
+#[derive(Debug)]
+pub(crate) struct Socket {
+    socket: UdpSocket,
+}
+
+impl Socket {
+    /// Transmit a packet if possible, with appropriate logging.
+    ///
+    /// We ignore I/O errors.  If a packet cannot be sent, we assume it is a transient condition
+    /// and drop it.  If it is not, the connection will eventually time out.  This provides a very
+    /// high degree of robustness.  Connections will transparently resume after a transient network
+    /// outage, and problems that are specific to one peer will not effect other peers.
+    pub(crate) fn poll_send_to(&self, cx: &mut Context<'_>, packet: &Transmit) -> Poll<Result<()>> {
+        match {
+            let fut = self.socket.send_to(&packet.contents, &packet.destination);
+            futures::pin_mut!(fut);
+            fut.poll(cx)
+        } {
+            Poll::Pending => Poll::Pending,
+            Poll::Ready(Ok(e)) => {
+                trace!("sent packet of length {} to {}", e, packet.destination);
+                Poll::Ready(Ok(()))
+            }
+            Poll::Ready(Err(e)) => {
+                warn!("Ignoring I/O error on transmit: {:?}", e);
+                Poll::Ready(Ok(()))
+            }
+        }
+    }
+
+    /// A wrapper around `recv_from` that handles ECONNRESET and logging
+    pub(crate) fn recv_from(
+        &self,
+        cx: &mut Context<'_>,
+        buf: &mut [u8],
+    ) -> Poll<Result<(usize, std::net::SocketAddr)>> {
+        loop {
+            match {
+                let fut = self.socket.recv_from(buf);
+                futures::pin_mut!(fut);
+                fut.poll(cx)
+            } {
+                Poll::Pending => {
+                    break Poll::Pending;
+                }
+                Poll::Ready(Ok(e)) => {
+                    trace!("received packet of length {} from {}", e.0, e.1);
+                    break Poll::Ready(Ok(e));
+                }
+                // May be injected by a malicious ICMP packet
+                Poll::Ready(Err(e)) if e.kind() == std::io::ErrorKind::ConnectionReset => {
+                    warn!("Got ECONNRESET from recv_from() - possible MITM attack")
+                }
+                Poll::Ready(Err(e)) => {
+                    warn!("Fatal I/O error: {:?}", e);
+                    break Poll::Ready(Err(e));
+                }
+            }
+        }
+    }
+}
+
+impl Socket {
+    pub(crate) fn new(socket: UdpSocket) -> Self {
+        Self { socket }
+    }
+}
+
+impl Pending {
+    pub(crate) fn send_packet(
```

Comment:
I am a bit confused by the `send_packet` function. Why can a pending packet send itself? In addition why does the `send_packet` function not only send itself, but also additional packets that are passed in through the closure?

Wouldn't it make more sense for this function to be a method of `Socket`? The first part would remain, the second part would live on the caller side calling `send_packet` in a loop.


First part:
```rust
 match socket.poll_send_to(cx, &transmit) {
                Poll::Pending => return Poll::Pending,
                Poll::Ready(Ok(())) => {}
                Poll::Ready(Err(e)) => return Poll::Ready(Err(e)),
            }
```

Second part:
```rust
        while let Some(transmit) = source() {
            trace!("trying to send packet!");
            match socket.poll_send_to(cx, &transmit) {
                Poll::Ready(Ok(())) => {}
                Poll::Pending => {
                    self.pending = Some(transmit);
                    return Poll::Pending;
                }
                Poll::Ready(Err(e)) => return Poll::Ready(Err(e)),
            }
        }
```

Reply:
> I am a bit confused by the `send_packet` function. Why can a pending packet send itself? In addition why does the `send_packet` function not only send itself, but also additional packets that are passed in through the closure?

This is the first asynchronous Rust code I ever wrote, so I had to learn as I went along.  As such, especially the older code in this PR is neither clean nor idiomatic.

> Wouldn't it make more sense for this function to be a method of `Socket`? The first part would remain, the second part would live on the caller side calling `send_packet` in a loop.

Indeed it would, provided that doing so does not introduce overhead.

Reply:
So this code is basically trying to enforce an invariant: whenever a packet can be sent it is.